python/ray/tests/test_gcs_ha_e2e.py

```diff
@@ -0,0 +1,190 @@
+import pytest
+import sys
+import threading
+from time import sleep
+
+from pytest_docker_tools import container, fetch, network
+from pytest_docker_tools import wrappers
+from http.client import HTTPConnection
+
+
+class Container(wrappers.Container):
+    def ready(self):
+        self._container.reload()
+        if self.status == "exited":
+            from pytest_docker_tools.exceptions import ContainerFailed
+
+            raise ContainerFailed(
+                self,
+                f"Container {self.name} has already exited before "
+                "we noticed it was ready",
+            )
+
+        if self.status != "running":
+            return False
+
+        networks = self._container.attrs["NetworkSettings"]["Networks"]
+        for (_, n) in networks.items():
+            if not n["IPAddress"]:
+                return False
+
+        if "Ray runtime started" in super().logs():
+            return True
+        return False
+
+    def client(self):
+        port = self.ports["8000/tcp"][0]
+        return HTTPConnection(f"localhost:{port}")
+
+
+gcs_network = network(driver="bridge")
+
+redis_image = fetch(repository="redis:latest")
+
+redis = container(
+    image="{redis_image.id}",
+    network="{gcs_network.name}",
+    command=(
+        "redis-server --save 60 1 --loglevel" " warning --requirepass 5241590000000000"
+    ),
+)
+
+header_node = container(
+    image="ray_ci:v1",
+    name="gcs",
+    network="{gcs_network.name}",
+    command=["ray", "start", "--head", "--block", "--num-cpus", "0"],
+    environment={"RAY_REDIS_ADDRESS": "{redis.ips.primary}:6379"},
+    wrapper_class=Container,
+    ports={
+        "8000/tcp": None,
+    },
+)
+
+worker_node = container(
+    image="ray_ci:v1",
+    network="{gcs_network.name}",
+    command=["ray", "start", "--address", "gcs:6379", "--block"],
+    environment={"RAY_REDIS_ADDRESS": "{redis.ips.primary}:6379"},
+    wrapper_class=Container,
+    ports={
+        "8000/tcp": None,
+    },
+)
+
+
+@pytest.fixture
+def docker_cluster(header_node, worker_node):
+    yield (header_node, worker_node)
+
+
+scripts = """
+import ray
```

Comment:
nit: is it possible to put the script into a separate python file and load it into here?